### ISSUE TITLE
Question/Doc request: smaller level of composition?

### ISSUE THREAD

**Issue by sthzg** [newtriks/generator-react-webpack#156]
First: thanks to all contributers for this generator. I've been using it for quite some time now and it is still my favorite boilerplate for react development.

I see why support for various technologies got dropped and am trying to understand the ways to "compose own features into the generator". I was looking at the yeoman docs and the implementation in [generator-react-webpack-alt](https://github.com/weblogixx/generator-react-webpack-alt). If I am not wrong this way of composition simply means writing your own generator and referencing base generators, which seems great for adding a reasonably large feature set (but also introduces all the restrictions of a *single inheritance* strategy).

So I am wondering if there are smaller levels of composition or more accurately simpler ways of **integrating custom features into the existing generator**. E.g. if I want to provide templates for flux things without maintaining a completely different generator or if I want to provide some flags that could alter existing templates similar to the way it is now supported with the `--stateless` option on the component template. I guess something like a "plugin-system" that 

a) does not have to live in this codebase
b) can be simply activated on my behalf

I note that the yeoman [docs on composition](http://yeoman.io/authoring/composability.html) mention some new ways so that an "end user may also initiate the composition". This seems to be WIP and there isn't much information about what it is (and what not) right now. As I didn't find more information about it I just know that it sounds like it could address that concern. :)

**TLDR;**
Do you know about other ways to integrate own things into an existing generator and if so, do you think it's worthwhile to add some documentation about it (could also be a Github wiki here)? Or maybe I am making it much too complicated and just don't know a common best practice to do this.

**Comment by weblogixx** [comment]
@sthzg: First thank you for your last contributions. I enjoy them very much. I think I am able to elaborate how this whole thing came to life and why it is done the way it is currently. Maybe this will answer some questions.

As I was starting out with React and Webpack, there where simply not that much useable projects to get you going. The one project that I found and that was worth a deeper look was generator-react-webpack, in this time in its 1.x release. Like you, I really liked the way it worked and I also liked the whole Flux-Implementation-Stuff that was in there. It just made it very easy to get started.

Because of those reasons, I used it some time and finally got to the conclusion that there where some features missing. However, @newtriks, the original author of the software (and still the current project maintainer) was very busy with other things and could not put as much time in the generator as he wanted.

Then came some updates for Flux implementations that could just not be tracked fast enough. It was also heavy to support all those flux frameworks, because no one knew them all. You cant be expert for everything, even if you try as hard as you can. That was the moment when we choose to strip the generator down to its bare bones. I still thank @newtriks that he gave me this chance, as I was not that active in the node community before filing my first PRs in here.

We initially tried to just revert the changes, but much of the older generator code was not that easy to maintain anymore. We knew the people would still love to use those flux implementations.

When I thought about how to we could to it, I also stumbled above the yeoman doc link you stated above. So I decided to break the generator down and use the afromentioned composition feature for those flux implementations.

When version 2.0 was ready, I directly started with the -alt version. There I found out that it works a bit different than what I expected it to work. Like you, I thought I could just add the -alt features more or less "on top" of the current generator. As yeoman is file based, this simply does not work. The composition feature is, as far as I understand it (and I found nothing on the web stating otherwise) ___not___ meant to enhance a given generator with features, but just to use the features of a given generator to enhance YOUR newly build generator.

That is also the reason you got the component subgenerator in generator-react-webpack-alt. It is just not possible to export the original interface without adding a (nearly) empty generator that composes the build in react-webpack function :(.

This means you always seem to need a small amount of boilerplate code. It also means that when I would choose to add a new subgenerator to generator-react-webpack, ALL generators using it as a base should also include this one. If you are not, you would have a system that works like this:

```
# The first steps are identical:
npm install -g yo
npm install -g generator-react-webpack
npm install -g generator-react-webpack-alt

cd /somewhere

# Version one: Using the current interface (speek, how it IS):
yo react-webpack-alt
yo react-webpack-alt:component newcomponent
yo react-wbpack-alt:component newstate --stateless
yo react-webpack-alt:store mystore

# How it would be otherwise:
yo react-webpack
yo react-webpack:component newcomponent
yo react-webpack:component newstate --stateless
yo react-webpack-alt:store mystore
```

I honestly just dont know if the second version (without the composition feature) would not be too difficult to use. Therefore, I opted to go the more "expensive" way to compose the interface of react-webpack.

I would also love to have it the different way around, but just dont see a possibility with the current yeoman system. As the yeoman authors have stated in the documentation, the composition feature is still in early stages. So maybe one time it will be possible to enhance a generator with plugins in a "valid" way.

Hope this helps,
Chris

**Comment by sthzg** [comment]
Thanks for your reply! Makes perfect sense. Like you said there are quite a few flux-implementations, all of them moving constantly and on top of it Facebook's Relay project that also influences the role of flux. Not to mention the choice of developing with or without flow. So yes, supporting that all in one monolithic generator wouldn't be possible.

I continued searching a bit for the second approach briefly mentioned in the Yeoman docs (end user initiating composition) but still haven't found something. I upped the [question on SO](http://stackoverflow.com/questions/33291455/what-does-composition-initiated-by-an-end-user-mean), maybe somebody will clarify there (also, a first read through the yeoman code also didn't reveal something, but I need to take a closer look).

Until there is something more generic I thought about another approach to provide custom subgenerators for an existing project. The basic idea is to provide a small command line script that can be added to the npm run scripts. It would accept two arguments, 

1. the URL to a ZIP containing a subgenerator
1. the name of the generator that this should be added to

The script simply downloads the zip, expands the relevant parts and injects it into the existing generator. To provide production-grade quality of such a tool involves quite a few aspects (like providing update/uninstall commands and a version check that can warn users when an update of the generator targets a later version of React as the subgenerator states to be compatible with).

If that sounds interesting I've started a [repository](https://github.com/sthzg/yo-subgen-installer/tree/feat_proof_of_concept_branch) where I am trying to come up with some proof-of-concept code. Not sure if it will work out though. ;-). Would be interested in your opinion on that approach (the code itself is purely prototyping code, doesn't guard against anything yet).

**Comment by sthzg** [comment]
@weblogixx if you are okay with it I vote for closing this as an issue, as there are not any actions implied and the initial question is answered. As also the devs from Yeoman have clarified that the "user initiated composition" is a planned feature (and nothing related to it is included in the code by now) there doesn't seem to be an official and elegant way to compose on a smaller scale.

**Comment by weblogixx** [comment]
@sthzg: I am watching your growing repo. Seems nice so far. But yes, I really would like to wait for something "official" from the Yeoman-guys before implementing something like this. I will keep an eye on it (and on your repo, too).
